zerver/tests/test_auth_backends.py

```diff
@@ -1627,6 +1610,42 @@ class GitHubAuthBackendTest(SocialAuthBase):
     AUTH_FINISH_URL = "/complete/github/"
     CONFIG_ERROR_URL = "/config-error/github"
 
+    def social_auth_test_finish(self, result: HttpResponse,
+                                account_data_dict: Dict[str, str],
+                                headers: Dict[str, Any]={},
+                                expect_choose_email_screen: bool=False,
+                                email_data: Optional[List[Dict[str, str]]]=None,
+                                **extra_data: Any) -> HttpResponse:
+        parsed_url = urllib.parse.urlparse(result.url)
+        csrf_state = urllib.parse.parse_qs(parsed_url.query)['state']
+        result = self.client_get(self.AUTH_FINISH_URL,
+                                 dict(state=csrf_state), **headers)
+        if email_data is None:
+            email_data = [account_data_dict]
```

Comment:
This `email_data` hack struck me as wrong; and I confirmed it was wrong when I added a bit more validation below.

The updated test uses `self.email_data`, which does what we need here and allows the more extensive validation below to be correct.